### ISSUE TITLE
[Feature] Batched asynchronous Engram host-row prefetch

### ISSUE THREAD

**Issue by zhangxiaolei123456** [sgl-project/sglang#38856]
### Checklist

- [ ] If this is not a feature request but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [ ] Please use English. Otherwise, it will be closed.

## Motivation

DeepSeek-V4.1 introduces two large, immutable Engram tables. SGLang can keep these tables in host memory with:

```bash
SGLANG_ENABLE_DSV41_ENGRAM_HOST_TABLE=1
```

This reduces HBM usage substantially, but every Engram lookup then performs sparse reads from pinned host memory. The current optional prefetch path:

```bash
SGLANG_ENABLE_DSV41_ENGRAM_KV_PREFETCH=1
```

launches the layer-14 Engram lookup and WKV projection on a side CUDA stream and overlaps them with layers 0-13. However, it is currently enabled only when all of the following are true:

- CUDA execution;
- pipeline parallel size is 1;
- DP attention is disabled;
- the model is text-only;
- the host table uses the shared layout;
- the WKV projection uses the FlashInfer CuTeDSL MXFP8 backend;
- the forward is normal decode with exactly one token.

The FlashInfer CuTeDSL MXFP8 backend requires Blackwell. H20 is SM90 and uses the block-FP8 fallback for the DeepSeek-V4.1 32x32 dense weights, so the final backend check prevents prefetch from being enabled even when the host table is shared and pinned.

For batch sizes greater than one, simply removing the `hash_ids.shape[0] == 1` check is undesirable. The layer-14 WKV projection is a large dense GEMM. Running it concurrently with layers 0-13 can compete for SM and HBM resources and may reduce throughput. On the other hand, the sparse host-row lookup uses little GPU compute and is a better candidate for overlap.

The requested feature is therefore a batched, rows-only prefetch path for DeepSeek-V4.1 host Engram tables on Hopper, with an optional per-GPU hot-row cache inspired by HugeCTR's GPU embedding cache.

baseline: #38798 

## Existing execution path

For each Engram layer, the current synchronous path is:

```text
hash IDs
  -> gather quantized rows from device or mapped host memory
  -> dequantize rows to BF16
  -> WKV projection
  -> Engram gate
```

For the existing batch-size-one prefetch path, layer 14 performs:

```text
side stream:  host gather -> WKV projection
main stream:  layers 0-13
layer 14:     wait for side stream -> gate
```

Relevant code:

- `python/sglang/srt/models/deepseek_v4.py`
  - construction of `engram_prefetch_stream`;
  - launch of `Engram.project`;
  - synchronization before layer 14.
- `python/sglang/srt/layers/engram.py`
  - `EngramEmbedding.forward`;
  - `Engram.project`;
  - shared pinned host-table implementation.
- `python/sglang/kernels/ops/embeddings/engram_gather.py`
  - GPU gather and dequantization from device or mapped host pointers.

## Goals

1. Hide layer-14 host Engram lookup latency for normal decode batches larger than one on H20 and other Hopper GPUs.
2. Keep the layer-14 WKV GEMM on the main stream for medium and large batches.
3. Preserve the existing complete lookup-plus-WKV prefetch for batch size one.
4. Preserve bitwise output equivalence with the synchronous path.
5. Keep all decisions and miss handling on the GPU without synchronizing the scheduler or worker thread.
6. Remain compatible with decode CUDA Graph capture and replay.
7. Allow an optional bounded GPU cache for frequently accessed Engram rows.
8. Introduce no behavior or memory change when the feature is disabled.

## Non-goals

- Changing the DeepSeek-V4.1 model definition or Engram hashing.
- Prefetching the layer-1 WKV projection.
- Running large WKV GEMMs concurrently with the main model stream.
- Supporting DP attention, pipeline parallelism, or vision in the first implementation.
- Supporting `TARGET_VERIFY`/DSpark in the first implementation.
- Adding mutable embedding updates or a distributed parameter server.
- Introducing HugeCTR as a runtime dependency.

### 1. Split lookup from projection

Split the current `Engram.project()` operation into two explicit stages:

```python
class Engram(nn.Module):
    def lookup(
        self,
        hash_ids: torch.Tensor,
        *,
        out: torch.Tensor | None = None,
    ) -> torch.Tensor:
        ...

    def project_embeddings(self, embeddings: torch.Tensor) -> torch.Tensor:
        ...

    def project(self, hash_ids: torch.Tensor) -> torch.Tensor:
        return self.project_embeddings(self.lookup(hash_ids))
```

`EngramEmbedding.forward()` and `engram_gather()` should accept an optional preallocated output tensor. This avoids allocations during CUDA Graph capture and gives the prefetch path stable buffer addresses.

The existing synchronous call sites continue to use `Engram.forward()` or `Engram.project()` and retain their current behavior.

### 2. Launch rows-only prefetch after layer 1

For medium decode batches, launch the layer-14 host-row lookup after layer 1 has finished its own Engram access:

```text
main stream:
  layer 0
  layer 1 host lookup -> WKV -> gate
  record layer-1-complete event
  layers 2-13
  wait for layer-14-rows-ready event
  layer 14 WKV -> gate

prefetch stream:
  wait for layer-1-complete event
  layer 14 host lookup -> BF16 staging buffer
  record layer-14-rows-ready event
```

Launching after layer 1 avoids concurrent random host-memory traffic from both Engram layers. Layers 2-13 still provide substantial latency-hiding distance.

The WKV projection remains on the main stream. This is the important difference from the existing batch-size-one path.

### 3. Prefetch policy

Use the number of local model-input rows, not the logical request count. This also provides the correct quantity if speculative verify is supported later.

Suggested policy:

```text
normal decode, 1 token:
    full prefetch: host lookup + WKV

normal decode, 2..full_prefetch_max_tokens:
    full prefetch only if the WKV backend explicitly reports concurrent-stream safety

normal decode, up to rows_prefetch_max_tokens:
    rows-only prefetch

larger batches:
    synchronous fallback
```

Initial defaults should be conservative:

```text
full_prefetch_max_tokens = 1
rows_prefetch_max_tokens = 64
```

The thresholds should remain opt-in until H20 benchmarks show a consistent benefit.

### 4. Explicit stream-safety capability

Do not infer stream safety from the GPU architecture or inspect backend fields with `getattr`. Add an explicit interface to the linear method abstraction:

```python
class LinearMethodBase:
    def supports_concurrent_cuda_stream(self) -> bool:
        return False
```

Backends may override this after they have been tested. Initially:

- FlashInfer CuTeDSL MXFP8: `True`, preserving current behavior.
- Hopper block-FP8 Triton: keep WKV on the main stream; no capability is needed for rows-only prefetch.
- Hopper DeepGEMM or FlashInfer CUTLASS: `False` until workspace ownership and concurrent execution are validated.

This prevents a future backend change from silently enabling an unsafe cross-stream GEMM.

### 5. CUDA stream and lifetime rules

The prefetch stream must be created during model initialization.

For every prefetch:

1. The prefetch stream waits for the main stream after hash IDs and layer-1 Engram work are ready.
2. The prefetch stream writes only to its dedicated staging buffer.
3. The main stream waits for the rows-ready event immediately before layer 14.
4. The staging tensor is recorded on the main stream before reuse.
5. Host-table mappings and model weights remain alive for the model lifetime.

No host-side `cudaStreamSynchronize`, `.item()`, or device-to-host copy is allowed in the request path.

### 6. CUDA Graph integration

CUDA Graph replay requires stable addresses and a deterministic branch for each captured shape.

- Allocate one staging buffer at the maximum configured prefetch token count, or one buffer per graph bucket.
- Slice the stable buffer for the current captured row count.
- Decide full, rows-only, or synchronous mode from the static captured token count.
- Capture the event dependencies between the main and prefetch streams.
- Do not allocate temporary tensors inside the captured prefetch path.
- Padded rows must produce the same ignored values as the current decode graph path.

The eager path can use the same buffers and policy.

### 7. Optional per-GPU hot-row cache

After rows-only prefetch is working, add an optional GPU cache inspired by HugeCTR's set-associative embedding cache.

Each GPU owns an independent cache for layer 14:

```text
GPU hash IDs
  -> GPU cache query
     -> hit: copy packed row from HBM
     -> miss: read packed row from mapped pinned host memory
              and optionally insert it into the cache
  -> dequantize into the BF16 staging buffer
```

The cache should store the original compressed representation:

```text
key:    Engram row ID
value:  FP8 payload + E8M0 scales
```

This preserves exact values and approximately doubles cache capacity compared
with storing BF16 rows.

Recommended initial implementation:

- separate cache per Engram layer;
- only instantiate the layer-14 cache;
- 2-way or 4-way set associativity;
- approximate clock/LRU replacement;
- configurable capacity, default disabled;
- query, miss fill, and replacement on the same prefetch stream;
- no CPU-visible missing count;
- no cross-GPU cache coherence, because the model table is immutable.

Unlike HugeCTR, SGLang does not need update handling, database backends, or thread-safe concurrent mutation. A single ordered prefetch stream permits a much smaller implementation.

### 8. Configuration

Keep the existing opt-in:

```bash
SGLANG_ENABLE_DSV41_ENGRAM_HOST_TABLE=1
SGLANG_ENABLE_DSV41_ENGRAM_KV_PREFETCH=1
SGLANG_DSV41_ENGRAM_HOST_TABLE_LAYOUT=shared
```

Add:

```bash
# auto: full for supported small shapes, rows-only for medium shapes
# rows: always rows-only within the token limit
# project: request full project prefetch and reject unsupported backends
SGLANG_DSV41_ENGRAM_PREFETCH_MODE=auto

# Number of local model-input rows, not logical requests.
SGLANG_DSV41_ENGRAM_PREFETCH_MAX_TOKENS=64

# Optional packed hot-row cache per GPU. Zero disables it.
SGLANG_DSV41_ENGRAM_GPU_CACHE_MB=0
```

The feature should log the resolved mode, backend, token limit, cache size, host
layout, pinning status, and the reason for any fallback.

### 9. Eligibility and fallback

The first implementation remains limited to:

- DeepSeek-V4.1 text checkpoints;
- CUDA;
- PP size 1;
- DP attention disabled;
- shared, pinned host Engram table;
- normal `DECODE` forward mode.

Unsupported combinations fall back to synchronous lookup. Explicit
`prefetch_mode=project` may raise a configuration error when the WKV backend is
not stream-safe; `auto` should fall back without failing the server.

`TARGET_VERIFY` must not be treated as an ordinary batch. DSpark uses
`batch_size * draft_token_num` model-input rows and separate graph shapes.
Support for it should be a follow-up after normal decode is validated.

## Performance model

Without prefetch:

```text
T = T_layers_0_13 + T_host_lookup_14 + T_wkv_14
```

With rows-only prefetch:

```text
T = max(T_layers_2_13, T_cache_query + T_host_misses_14) + T_wkv_14
```

Rows-only prefetch is beneficial when:

```text
T_cache_query + T_host_misses_14 < T_layers_2_13
```

and the side-stream gather does not materially slow the main stream.

For the published DeepSeek-V4.1 layout, each token accesses approximately 24 Engram rows. With a 256-element FP8 row and E8M0 scales, the packed host traffic is approximately 6.3 KiB per token and the BF16 staging output is approximately 12 KiB per token. At 64 tokens this is still less than 1 MiB of staging data. The expected bottleneck is sparse PCIe transaction and TLB latency rather than bulk bandwidth, making it suitable for overlap with layers 2-13.

The WKV projection is much larger, approximately:

```text
[tokens, 6144] x [25600, 6144]^T
```

It should remain on the main stream for medium and large batches unless
profiling proves concurrent execution beneficial.

## Correctness requirements

1. Rows-only prefetch must be bitwise equivalent to the synchronous host lookup followed by main-stream WKV.
2. A cache hit must return the exact FP8 payload and E8M0 scale bytes stored in the host table.
3. Cache misses and collisions must never return stale rows.
4. Padded CUDA Graph rows must not affect real outputs.
5. Cache state must not be visible across model instances or checkpoint reloads.
6. Disabling the feature must preserve current execution and memory behavior.
7. Failure to allocate the optional cache should produce a clear error or deterministic fallback, not partial enablement.

### Unit tests

- Refactored `lookup`, `project_embeddings`, and `project` equivalence.
- Prefetch-policy selection for token counts 1, 2, 4, 8, 16, 32, 64, and 65.
- Eligibility rejection for device tables, private host layout, DP attention, PP, vision, and unsupported forward modes.
- Cache hit, miss, collision, replacement, and empty-cache behavior.
- Exact packed FP8/E8M0 cache round trip.

### CUDA tests

- Synchronous versus rows-only output equality.
- Repeated side-stream execution without races.
- Decode CUDA Graph capture and replay for all configured buckets.
- Eager decode with changing batch sizes.
- Cache enabled and disabled.
- Host table pinned with shared layout.
- At least 1,000 repeated decode iterations under compute-sanitizer/racecheck where practical.

### End-to-end tests

- DeepSeek-V4.1 text generation on SM90.
- Batch sizes 1, 8, 16, 32, and 64.
- Radix-cache hit and miss cases.
- Mixed request completion where the active batch shrinks between decode steps.
- Output comparison against prefetch disabled.

## Benchmark plan

Benchmark on 8x H20 with the same model, request set, graph buckets, and memory
limits for all variants:

1. Host table, synchronous lookup.
2. Host table, rows-only prefetch.
3. Host table, rows-only prefetch plus GPU row cache.
4. Existing full-project prefetch at batch size one.

Measure:

- inter-token latency and output tokens/s;
- layer-14 prefetch duration;
- main-stream wait time at layer 14;
- layer 2-13 elapsed time with and without prefetch;
- PCIe receive throughput and transactions;
- SM and HBM utilization;
- cache hit rate and miss rows per token;
- graph capture memory and steady-state HBM usage.

Suggested workloads:

```text
decode batch: 1, 4, 8, 16, 32, 64
context:      1K, 8K, 32K, 128K
prompt type:  repeated-prefix, natural-language mix, code-agent mix, random tokens
```

The feature should remain opt-in unless:

- output correctness is unchanged;
- rows-only prefetch improves representative H20 batch 8-64 TPOT or throughput;
- no tested batch has more than 1% regression;
- the main stream shows no material slowdown from the prefetch kernel.

The GPU cache should remain disabled by default until production traces show
enough temporal locality to amortize cache lookup and replacement.
